platforms/chibios/interrupt_handlers.c

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Nick Brassel (@tzarc)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+///////////////////////////////////////////////////////////////////////////////
+// BEGIN: STM32L4xx Wear-leveling ECC fault handling
+//
+// STM32L4xx have ECC checks for all flash memory access. Whenever there's an
+// ECC failure, the MCU raises the NMI interrupt -- in ChibiOS this is normally
+// used for context switching, but in the case of L4xx we need to ensure the
+// method used for context switching does not collide with the NMI interrupt.
+//
+// Whenever we do receive an ECC-triggered NMI interrupt, we gracefully cater
+// for it, signalling the wear-leveling code that a failure has occurred.
+///////////////////////////////////////////////////////////////////////////////
+
+#include <ch.h>
+#include <chcore.h>
+
+#if defined(QMK_MCU_SERIES_STM32L4XX) && defined(WEAR_LEVELING_EMBEDDED_FLASH)
+
+extern bool backing_store_allow_ecc_errors(void);
+extern void backing_store_signal_ecc_error(void);
+
+void NMI_Handler(void) {
+#    ifdef FLASH_ECCR_ECCD
+    if (FLASH->ECCR & FLASH_ECCR_ECCD) {
+        if (backing_store_allow_ecc_errors()) {
+            FLASH->ECCR = FLASH_ECCR_ECCD;
+            backing_store_signal_ecc_error();
+            return;
+        }
+    }
+#    endif // FLASH_ECCR_ECCD
+
+    chSysHalt("NMI");
+}
```

Comment:
And yet another thing which would be required to move the vector table to RAM is `#define CRT0_VTOR_INIT TRUE`, but that's apparently the default for ARM>=v7.  The situation with ARMv6 is more complicated, because plain Cortex-M0 cores don't have the VTOR register at all (it's present only in Cortex-M0+).  On STM32 chips with Cortex-M0 cores the only way to move the vector table in RAM is to change some mapping bits to map SRAM at address 0 (apparently it is possible to do after boot).  Although we could just ignore that, because apparently only some newer chips actually have ECC flash which generates NMIs.

Reply:
I think I'm leaning towards removing the execute-from-flash requirement. The Keychron branch they've modified for it doesn't bother to relocate to RAM, and they were the original reporters of this problem. My guess is that unless the NMI interrupt occurs _again_ when executing the handler code, it's probably fine.